core/src/test/java/com/google/googlejavaformat/java/MainTest.java

```diff
@@ -613,4 +614,49 @@ public void noFormatJavadoc() throws Exception {
     assertThat(main.format("--skip-javadoc-formatting", "-")).isEqualTo(0);
     assertThat(out.toString()).isEqualTo(joiner.join(input));
   }
+
+  private static Path createTestSourceFile(TemporaryFolder testFolder, String className)
+      throws IOException {
+    Path path = testFolder.newFile(className + ".java").toPath();
+    Files.write(path, ("class " + className + " {}\n").getBytes(UTF_8));
+    try {
+      Files.setPosixFilePermissions(path, EnumSet.of(PosixFilePermission.OWNER_READ));
+    } catch (UnsupportedOperationException e) {
+      throw new IOException(e);
+    }
+    return path;
+  }
+
+  /**
+   * Reproduces https://github.com/google/google-java-format/issues/846: Formatter leaks threads and
+   * memory
+   */
+  @Test
+  public void doesNotLeakThreads() throws Exception {
+    Path testA = createTestSourceFile(testFolder, "TestA");
+    Path testB = createTestSourceFile(testFolder, "TestB");
+
+    // Get the number of threads before formatting.
+    int numberOfThreadsBeforeFormatting = Thread.currentThread().getThreadGroup().activeCount();
+
+    // Run the formatter several times. If you bump the number of iterations to ~10000, you can
+    // reproduce an OOM error.
+    for (int i = 0; i < 50; i++) {
+      int errorCode =
+          Main.main(
+              new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out, UTF_8)), true),
+              new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.err, UTF_8)), true),
+              "--replace",
+              testA.toString(),
+              testB.toString());
+      assertWithMessage("Error Code").that(errorCode).isEqualTo(0);
+    }
+
+    // Verify that the number of threads is nearly the same before and after. NB: we allow for a few
+    // additional threads since we don't completely control the number of running threads in this
+    // test. In the failing case there are ~100 extra threads, so this should be sufficient to catch
+    // bugs.
+    assertThat(Thread.currentThread().getThreadGroup().activeCount())
+        .isLessThan(numberOfThreadsBeforeFormatting + 5);
```

Comment:
Thanks for taking the time to write a test here. This seems hard to test in a unit test that's quick to run and non-flaky, though--I'm thinking of the comment about 10k iterations, and checking for the number of threads `+-5`. I think I'd be OK with just skipping a test for this.

Reply:
Removed.